versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java

```diff
@@ -1334,4 +1336,74 @@ protected void checkExpectedGlobalStates(
       }
     }
   }
+
+  /** Load the refLog entry for the given hash, return {@code null}, if not found. */
+  protected abstract RefLogEntry fetchFromRefLog(OP_CONTEXT ctx, Hash refLogId);
+
+  /**
+   * Fetch multiple {@link RefLogEntry refLog-entries} from the refLog. The returned list must have
+   * exactly as many elements as in the parameter {@code hashes}. Non-existing hashes are returned
+   * as {@code null}.
+   */
+  protected abstract List<RefLogEntry> fetchPageFromRefLog(
+      OP_CONTEXT ctx, List<RefLogEntry.Ancestor> ancestors);
+
+  /** Reads from the refLog starting at the given refLog-hash. */
+  protected Stream<RefLogEntry> readRefLogStream(OP_CONTEXT ctx, Hash initialHash)
+      throws RefLogNotFoundException {
+    Spliterator<RefLogEntry> split = readRefLog(ctx, initialHash);
+    return StreamSupport.stream(split, false);
+  }
+
+  protected Spliterator<RefLogEntry> readRefLog(OP_CONTEXT ctx, Hash initialHash)
+      throws RefLogNotFoundException {
+    if (NO_ANCESTOR.equals(initialHash)) {
+      return Spliterators.emptySpliterator();
+    }
+    RefLogEntry initial = fetchFromRefLog(ctx, initialHash);
+    if (initial == null) {
+      throw RefLogNotFoundException.forRefLogId(initialHash.asString());
+    }
+    return refLogFetcher(ctx, initial, this::fetchPageFromRefLog, RefLogEntry::getAncestors);
+  }
+
+  protected <T> Spliterator<T> refLogFetcher(
```

Comment:
This one should be unified with the existing method - and not copied.

Reply:
yes. done.